sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/TestLocationAnalyzerTest.cs

```diff
@@ -0,0 +1,407 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+extern alias csharp;
+extern alias vbnet;
+
+using System.Linq;
+using FluentAssertions;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SonarAnalyzer.Protobuf;
+using SonarAnalyzer.UnitTest.Helpers;
+using CSharpTestLocationAnalyzer = csharp.SonarAnalyzer.Rules.TestLocationAnalyzer;
+using VisualBasicTestLocationAnalyzer = vbnet.SonarAnalyzer.Rules.TestLocationAnalyzer;
+
+namespace SonarAnalyzer.UnitTest.Rules
+{
+    [TestClass]
+    public class TestLocationAnalyzerTest
+    {
+        [DataTestMethod]
+        [DataRow("1.1.11")]
+        [DataRow(Constants.NuGetLatestVersion)]
+        [TestCategory("Rule")]
+        public void FileMetadataAnalyzer_MsTest_CSharp(string mstestVersion)
+        {
+            var projectFiles = new[]
+            {
+                "TestCases/TestClassShouldHaveTestMethod.MsTest.cs",
+                "TestCases/TestMethodShouldHaveCorrectSignature.MsTest.cs",
+            };
+
+            Verifier.VerifyUtilityAnalyzer<TestLocation>(
+                projectFiles,
+                new CSharpTestLocationAnalyzer_
+                {
+                    IsEnabled = true,
+                    WorkingPath = @"TestLocationAnalyzer",
+                },
+                @"TestLocationAnalyzer\test-location.pb",
+                (messages) =>
+                {
+                    messages.Should().HaveCount(projectFiles.Length + 1); // + ExtraEmptyFile.g.cs
+
+                    messages[0].FilePath.Should().EndWith("ExtraEmptyFile.g.cs");
+                    messages[0].TestNames.Should().BeEmpty();
+
+                    messages[1].FilePath.Should().EndWith("TestClassShouldHaveTestMethod.MsTest.cs");
+                    messages[1].TestNames.Should().OnlyContain(
+                        "Tests.Diagnostics.ClassTest9.Foo",
+                        "Tests.Diagnostics.ClassTest12.Foo",
+                        "Tests.Diagnostics.Foo_WhenFoo_ExpectsFoo.Foo_WhenFoo_ExpectsFoo"
+                        );
+
+                    messages[2].FilePath.Should().EndWith("TestMethodShouldHaveCorrectSignature.MsTest.cs");
+                    messages[2].TestNames.Should().OnlyContain(
+                        "Tests.Diagnostics.MsTestTest.PrivateTestMethod",
+                        "Tests.Diagnostics.MsTestTest.ProtectedTestMethod",
+                        "Tests.Diagnostics.MsTestTest.InternalTestMethod",
+                        "Tests.Diagnostics.MsTestTest.AsyncTestMethod",
+                        "Tests.Diagnostics.MsTestTest.GenericTestMethod",
```

Comment:
We report all kinds of methods because XUnit can execute private, generic, etc. methods as tests... We don't distinguish between test frameworks and report all attributed methods.